src/mouse.c

```diff
@@ -0,0 +1,140 @@
+/*
+   Copyright (C) 2015 The University of Notre Dame
+   This software is distributed under the GNU General Public License.
+   See the file LICENSE for details.
+   */
+
+#include "mouse.h"
+#include "interrupt.h"
+#include "console.h"
+#include "ps2.h"
+#include "ioports.h"
+#include "process.h"
+#include "graphics.h"
+
+static uint8_t mouse_cycle = 0;
+static uint8_t mouse_byte[3];
+struct graphics_color mouse_fg_color = {0, 255, 0};
+
+static struct list queue = { 0, 0 };
+
+int mouse_scan() {
+    // see if data to be read on PS2_DATA_PORT
+    ps2_controller_read_ready();
+    int data = inb(PS2_DATA_PORT);
+    iowait();
+
+    return data;
+}
+
+// TODO: handle button events (e.g., downclick and release, doubleclick)
+void mouse_map() {
+    // signs for x and y movement (bytes 2 and 3); default to positive
+    bool x_sign = 0;
+    bool y_sign = 0;
+
+    // handle byte 1
+    int8_t byte1 = mouse_byte[0];
+    if ((byte1 & 0x01) == 1) {
+        // handle left button
+    }
+    if ((byte1 & 0x02) == 1) {
+        // handle right button
+    }
+    if ((byte1 & 0x04) == 1) {
+        // handle middle button
+    }
+    // mouse movement sign x
+    if ((byte1 & 0x10) == 1) {
+        x_sign = 1;
+    }
+    // mouse movement sign y
+    if ((byte1 & 0x20) == 1) {
+        y_sign = 1;
+    }
+
+    int8_t byte2 = mouse_byte[1];
+    int8_t byte3 = mouse_byte[2];
+
+    // neg sign bytes, if necessary
+    byte2 = x_sign ? (byte2 | 0xFFFFFF00) : byte2;
+    byte3 = y_sign ? (byte3 | 0xFFFFFF00) : byte3;
+    mouse_x += byte2;
+    // negative values are down with mouse offsets (opposite of graphics
+    mouse_y -= byte3;
+}
+
+int mouse_request_packet() {
+    // command to request a single packet
+    ps2_command_write(0xEB , PS2_COMMAND_REGISTER);
+    ps2_controller_read_ready();
+    int ack = inb(PS2_DATA_PORT);
+    if (ack != MOUSE_ACK) {
+        // TODO: handle ack
+        console_printf("mouse: bad ack: %x\n", ack);
+    }
+    ps2_controller_read_ready();
+    return inb(PS2_DATA_PORT);
+}
+
+void mouse_interrupt() {
+    // mice sends 3 bytes; only process on third byte
+    // TODO: handle 4th byte (scroll wheel)
+    switch(mouse_cycle) {
+        case 0:
+            mouse_byte[0] = inb(PS2_DATA_PORT);
+            mouse_cycle++;
+            break;
+        case 1:
+            mouse_byte[1] = inb(PS2_DATA_PORT);
+            mouse_cycle++;
+            break;
+        case 2:
+            mouse_byte[2] = inb(PS2_DATA_PORT);
+            mouse_cycle++;
+            mouse_cycle = 0;
+            mouse_map();
+            graphics_mouse();
+            break;
+    }
+
+    process_wakeup(&queue);
```

Comment:
Im still confused by this. You are providing an empty queue, to which you never add anything. What does this do?